waiter/src/waiter/instance_tracker.clj

```diff
@@ -0,0 +1,89 @@
+          query-state-fn
+          (fn event-maintainer-query-state-fn
+            [include-flags]
+            (let [{:keys [all-failed-instances last-instance-handler-error-time last-update-time]} @state-atom]
+              (cond-> {:last-instance-handler-error-time last-instance-handler-error-time
+                       :last-update-time last-update-time}
+                      (contains? include-flags "all-failed-instances")
+                      (assoc :all-failed-instances all-failed-instances))))
+          go-chan
+          (async/go
+            (try
+              (loop [{:keys [all-failed-instances] :as current-state} @state-atom]
+                (reset! state-atom current-state)
+                (let [[msg current-chan] (async/alts! [exit-chan error-chan router-state-chan query-chan] :priority true)
+                      next-state
+                      (condp = current-chan
+                        exit-chan
+                        (do
+                          (log/warn "stopping event-maintainer")
+                          (when (not= :exit msg)
+                            (throw (ex-info "Stopping event-maintainer" {:time (clock) :reason msg}))))
+
+                        error-chan
+                        (timers/start-stop-time!
+                          (metrics/waiter-timer "core" "event-maintainer" "error-chan")
+                          (let [{:keys [errors instance-event-type]} msg
+                                last-instance-handler-error-time (clock)]
+                            (log/error "generating new error time for instance event handler error"
+                                       {:errors errors
+                                        :instance-event-type instance-event-type
+                                        :last-instance-handler-error-time last-instance-handler-error-time})
+                            (assoc current-state :last-instance-handler-error-time last-instance-handler-error-time)))
+
+                        router-state-chan
+                        (timers/start-stop-time!
+                          (metrics/waiter-timer "core" "event-maintainer" "router-state-chan")
+                          (let [{:keys [service-id->failed-instances]} msg
+                                all-failed-instances' (reduce (fn [cur-set cur-failed-instances]
+                                                                (set/union cur-set (set cur-failed-instances)))
+                                                              #{}
+                                                              (vals service-id->failed-instances))
+                                new-failed-instances (filter (complement all-failed-instances) all-failed-instances')
+                                error-cb-fn! (fn error-cb! [errors]
+                                               (async/put! error-chan {:errors errors :instance-event-type :instance-failure}))]
```

Comment:
Not sure we need an `error-cb-fn!`, we can handle it directly here and update state.